File: avg/average.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

double average(double *data, size_t count) {
    double sum = 0.0;
    for (size_t i = 0; i < count; i++)
        sum += data[i];
    return sum / count;
}

int main(void) {
    const size_t N = 100000000;
    double *data = malloc(N * sizeof(double));
    if (data == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        return 1;
    }

    srand((unsigned)time(NULL));

    clock_t startCreate = clock();
    for (size_t i = 0; i < N; i++)
        data[i] = ((double)rand() / RAND_MAX) * 200.0 - 100.0;
    clock_t endCreate = clock();

    clock_t startAverage = clock();
    double avg = average(data, N);
    clock_t endAverage = clock();

    clock_t endTotal = clock();

    double createTime = (double)(endCreate - startCreate) / CLOCKS_PER_SEC;
    double averageTime = (double)(endAverage - startAverage) / CLOCKS_PER_SEC;
    double totalTime = (double)(endTotal - startCreate) / CLOCKS_PER_SEC;

    printf("average = %f\n", avg);
    printf("Data creation: %.6f seconds\n", createTime);
    printf("Averaging:     %.6f seconds\n", averageTime);
    printf("Total:         %.6f seconds\n", totalTime);

    free(data);
    return 0;
}

File: avg/average.zig
const std = @import("std");

fn average(data: []const f64) f64 {
    var sum: f64 = 0.0;
    var count: usize = 0;

    for (data) |v| {
        sum += v;
        count += 1;
    }

    return sum / @as(f64, @floatFromInt(count));
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const start_total = std.time.nanoTimestamp();

    const start_create = std.time.nanoTimestamp();
    const n = 100_000_000;
    var data = try std.ArrayList(f64).initCapacity(allocator, n);
    defer data.deinit(allocator);

    var prng = std.Random.DefaultPrng.init(blk: {
        var seed: u64 = undefined;
        try std.posix.getrandom(std.mem.asBytes(&seed));
        break :blk seed;
    });
    const rng = prng.random();

    var i: usize = 0;
    while (i < n) : (i += 1) {
        const val = rng.float(f64) * 200.0 - 100.0;  // range: -100 to 100
        data.appendAssumeCapacity(val);
    }
    const end_create = std.time.nanoTimestamp();

    const start_average = std.time.nanoTimestamp();
    const avg = average(data.items);
    const end_average = std.time.nanoTimestamp();

    const end_total = std.time.nanoTimestamp();

    const create_time = @as(f64, @floatFromInt(end_create - start_create)) / 1_000_000_000.0;
    const average_time = @as(f64, @floatFromInt(end_average - start_average)) / 1_000_000_000.0;
    const total_time = @as(f64, @floatFromInt(end_total - start_total)) / 1_000_000_000.0;

    std.debug.print("average = {d}\n", .{avg});
    std.debug.print("Data creation: {d:.6} seconds\n", .{create_time});
    std.debug.print("Averaging:     {d:.6} seconds\n", .{average_time});
    std.debug.print("Total:         {d:.6} seconds\n", .{total_time});
}
